### PR TITLE
[SPARK-35449][SQL] Only extract common expressions from CaseWhen values if elseValue is set

### PR DESCRIPTION
**Pull request by Kimahriman**
### What changes were proposed in this pull request?

This PR fixes a bug with subexpression elimination for CaseWhen statements. https://github.com/apache/spark/pull/30245 added support for creating subexpressions that are present in all branches of conditional statements. However, for a statement to be in "all branches" of a CaseWhen statement, it must also be in the elseValue.

### Why are the changes needed?

Fix a bug where a subexpression can be created and run for branches of a conditional that don't pass. This can cause issues especially with a UDF in a branch that gets executed assuming the condition is true.

### Does this PR introduce _any_ user-facing change?

Yes, fixes a potential bug where a UDF could be eagerly executed even though it might expect to have already passed some form of validation. For example:
```
val col = when($"id" < 0, myUdf($"id"))
spark.range(1).select(when(col > 0, col)).show()
```

`myUdf($"id")` is considered a subexpression and eagerly evaluated, because it is pulled out as a common expression from both executions of the when clause, but if `id >= 0` it should never actually be run.

### How was this patch tested?

Updated existing test with new case.

### CODE REVIEW

**Comment by Kimahriman** [comment]
@viirya

**Comment by viirya** [comment]
ok to test

**Review by viirya — COMMENTED** [review]
looks good. waiting for CI

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
cc @maropu @cloud-fan too

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:230]: `conditions4` is not used anywhere.

**Comment by cloud-fan** [comment]
is it a perf-only issue?

**Comment by viirya** [comment]
> is it a perf-only issue?

In most cases, it is. But what I understand from @Kimahriman's description, seems in their usage like `when(..., udf(col))`, `udf(col)` will be wrongly treated as sub-expr and run in advance. But the udf fails for the input because it doesn't satisfy the condition.

So it could cause query error.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala
Comment (Kimahriman) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:230]: Copy pasta, fixed the below line

**Comment by Kimahriman** [comment]
> is it a perf-only issue?

Yeah because of the UDF issue I'd consider it more a bug with performance side-effects. Whether those side-effects are positive or negative largely depends on whether https://github.com/apache/spark/pull/32559 is merged. Without it, this can increase performance by reducing the cases where you could have unused subexpressions generated. With it, it can decrease performance by not being able to create subexpressions for simple when clauses like `when(length(col) > 0, col)`

**Comment by HyukjinKwon** [comment]
@Kimahriman would you mind describing what behaviour change (bug fix) happens in "Does this PR introduce any user-facing change?"? The fix itself looks making sense but it would be great to clarify what bug is fixed by this too.

**Comment by Kimahriman** [comment]
> @Kimahriman would you mind describing what behaviour change (bug fix) happens in "Does this PR introduce any user-facing change?"? The fix itself looks making sense but it would be great to clarify what bug is fixed by this too.

Updated, sorry I never know what to put for that section for bug fixes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:131]: > However, for a statement to be in "all branches" of a CaseWhen statement, it must also be in the elseValue.

It's better to leave a simple comment here about why we need to check `elseValue`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:239]: Could you put this new test in a new test unit `test("SPARK-35449: ..."?

**Review by maropu — COMMENTED** [review]
The fix itself looks fine. I left minor comments. btw, could you fill the description in the jira? It looks empty now: https://issues.apache.org/jira/browse/SPARK-35449

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala
Comment (Kimahriman) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala]: Moved to a new test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala
Comment (Kimahriman) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:148]: Added a comment in

**Comment by Kimahriman** [comment]
> The fix itself looks fine. I left minor comments. btw, could you fill the description in the jira? It looks empty now: https://issues.apache.org/jira/browse/SPARK-35449

I updated the JIRA and rebased off master to get the latest subexpression PRs in. Had to update one test based on these changes.

**Review by maropu — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by viirya** [comment]
Thanks all! Merging to master/3.1. I guess there will be conflict on 3.1, but let me try first.

**Comment by viirya** [comment]
This should to be fixed in branch-3.1 too. @Kimahriman Can you make a backport PR?

Also cc @dongjoon-hyun as he is release manager of 3.1.2.

**Comment by viirya** [comment]
Ah, right, @maropu, because backport to 3.1 fails, so the script doesn't update it. I forgot to do it manually too. Let me do it now. Thanks!

**Comment by Kimahriman** [comment]
> This should to be fixed in branch-3.1 too. @Kimahriman Can you make a backport PR?
> 
> Also cc @dongjoon-hyun as he is release manager of 3.1.2.

Created https://github.com/apache/spark/pull/32651

**Pull request closed without merging**
